plugins/nginx-vhosts/command-functions

```diff
@@ -44,7 +44,7 @@ cmd-nginx-report-single() {
     "--nginx-bind-address-ipv6: $(fn-plugin-property-get-default "nginx" "$APP" "bind-address-ipv6" "::")"
     "--nginx-disable-custom-config: $(fn-plugin-property-get-default "nginx" "$APP" "disable-custom-config" "false")"
     "--nginx-error-log-path: $(fn-nginx-error-log-path "$APP")"
-    "--nginx-hsts: $(fn-plugin-property-get-default "nginx" "$APP" "hsts" "true")"
+    "--nginx-hsts: $(fn-plugin-property-get-default "nginx" "$APP" "hsts" "$(fn-plugin-property-get-default "nginx" "--global" "hsts" "true")")"
```

Comment:
No, we should have separate report outputs for global vs app specific. So `--nginx-hsts` and `--nginx-global-hsts`. No need to change the default for the app-specific one.

Reply:
hm.. yeah ok.. but the app specific default is dependent on the global

Reply:
so if the global is false then the app default needs to be false

Reply:
````shell
    "--nginx-global-hsts: $(fn-plugin-property-get-default "nginx" "--global" "hsts" "true")"
    "--nginx-hsts: $(fn-plugin-property-get-default "nginx" "$APP" "hsts" "$(fn-plugin-property-get-default "nginx" "--global" "hsts" "true")")"
````

Reply:
else it is confusing.... global is false but app is true.. but as app has no own setting it is actually false.. so should display false

Reply:
Right but this is the report. Downstream tools can interpret the two values to figure out what to do.

If anything, we should change how the value gets pulled internally - maybe a helper function to fetch the global value with fallback to the app one? - and then report the values with empty default for app-specific and `true` default for global.

Reply:
yeah but if the default is false and the app has never explicitly set a value it should be false.. a tool would not know this

i.e. if an app has never enabled hsts it's default value is true, but the global value is false. So the app should display false. If an app actually sets it as true then it is true..?

Reply:
Right but again, this is the report. Lets say someone wants to force set it to true and has it globally set to true. The report wouldn't give the actual value of the property, but a computed one.

You'll see something similar with the resource properties that can have defaults set for all process types within an app. We show the values separately, not in computed fashion.

Reply:
hm ok, still as a user/dev I find that confusing.. the report says that the global property is disabled, but the app property is enabled. Even though (as the property has never been set) the app is actually disabled...

We seem to be speaking in circles ;-)

Out of the box dokku the app hsts default property is true
Setting hsts property globally to false gives you a report that says that globally it is false but the app is true (as it is the default)
This makes it confusing, as my report tells me that it is disabled but enabled for this app; which is not true

Reply:
What I'm saying is that we remove the app default, so `true` => "" (empty string). A user would need to parse that in their head.

The default global value would be true.

Reply:
Maybe I do not understand what you mean with report... I assume the current active config/props... something that displays what I want to see to understand how/what is active..

Reply:
If the app is forced to true then it has an entry? so would not go to the default; which is the global check

Reply:
So this particular diff comment is on a piece of code that gets run when someone calls `nginx:report` (or `report $APP`).

We want to expose the properties set by the user so that either automation can set the values as expected, or that they can report issues and I can debug what the actual state of their system is without asking more questions.

In that case, we should report the values _as set by the user_ unless there are system defaults. If we add a global default, then the app-specific version has no actual defaults, and should be an empty string.

Outside of this usage, we should probably have a single function - `fn-nginx-vhosts-hsts-enabled` maybe? - that fetches the app-specific value (with a default of empty string). If the fetched value is not empty, then it is returned. Otherwise, we fetch the global property value (with a default of `true`) and return that. This function would be used in every place outside of the report function to fetch whether or not hsts is enabled.

Reply:
so if global is set and app has never been set it should be empty for the app in the report?

could you otherwise post an example of how you would like the report to look like in the following situations:

 * out of the box app hsts
 * out of the box app hsts, with global hsts disabled
 * out of the box app hsts, with global hsts disabled, app hsts enabled
 * out of the box app hsts, with global hsts disabled, app hsts disabled

I think those are the cases

The report itself is also up to discussion as if I have disabled hsts on an app why does the report display hsts things that are not useful.. ; if disabled do not display parameters; disabled is disabled

Reply:
```
    "--nginx-global-hsts: $(fn-plugin-property-get-default "nginx" "--global" "hsts" "true")"
    "--nginx-hsts: $(fn-plugin-property-get-default "nginx" "$APP" "hsts" "")"
```

Those are the two lines that should be in the report function diff.

Reply:
The app-specific report value for nginx-hsts would have an empty default.

A similar condition occurs for the `deploy-branch` in the git plugin (and actually the output there should be default empty string, so thats wrong and will be fixed in a minor release).

The resource plugin is also similar in that you can set a default value - `_default_` - and that is shown separately from the explicitly set value.

Reply:
ah but you are missing the point the app hsts default is true

.. so even without this change the app hsts default displayed value must be true

current:
`"--nginx-hsts: $(fn-plugin-property-get-default "nginx" "$APP" "hsts" "true")"`

Reply:
else it has no meaning;

if hsts global false the app must display a default (if not set) that is correct and understandable; i.e. false or true

Reply:
I understand your point. From your vantage point, it is nice to know the computed value vs the actual. For me - and for automation purposes - the computed value is not interesting.

If you want, adding a new flag - `--nginx-hsts-is-enabled` or something - would be fine. Not all properties can be set - there's definitely a bunch of precedent for that - and this allows us to have the computed value show up in addition to setable ones.

Reply:
anyway semantics.

If you want what you are describing that is fine.

As a user I would like a report that actually describes the state; a readable report.. not something that I would have to dig in to the config/props to find out what the state actually is.

Reply:
But thats the thing, this PR is effectively saying that the app specific value has no default and falls back to a global value - which is already a BC-break. The state of the system is actually changed by this PR, as the default interpolated value is no longer actually correct.

Using a computed value - where we fallback to the global - as the interpolated value would be wrong from an automation sense:

- Globally `true`, app-specific has no user-set value: If I skip setting the app-specific value to `true` in my automation because it shows as `true`, and my next step is to set it globally to `false`, then the system is actually out of sync. Needing to remember order of operations is bad

For a report sense, I need to know if the user has the property set in one way or another so I can effectively debug what they have setup.

If we want to have something that can be inferred by the user, then adding a new computed report value is the way to go.

One thing that might make this more clear is if we had different prefixes for setable properties, so `--nginx-prop-hsts` would be the exact property setting, while `--nginx-hsts-is-enabled` would be the user-friendly report value.